src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -576,20 +575,26 @@ export class SearchSource {
 
       // request the remaining fields from stored_fields just in case, since the
       // fields API does not handle stored fields
-      const remainingFields = difference(uniqFieldNames, [
+      let remainingFields = difference(uniqFieldNames, [
         ...Object.keys(body.script_fields),
         ...Object.keys(body.runtime_mappings),
       ]).filter(Boolean);
 
-      // only include unique values
-      body.stored_fields = [...new Set(remainingFields)];
+      if (body._source && body._source.excludes) {
+        // TODO: optimize complexity
+        remainingFields = remainingFields.filter((remainingField) => {
+          return !body._source.excludes.includes(remainingField);
+        });
+      }
```

Comment:
We already have `fieldWildcardFilter` above which does this, maybe that can be reused, or this filtering can be done earlier on?

Or actually: It seems like if we know we _always_ want to filter items from `excludes`, we should filter `body.script_fields` and `body.runtime_mappings` as early as possible, so those values are never there to begin with?

Reply:
Agreed, but we'd still need to explicitly exclude `fieldsFromSource`, right? So we'd end up with the same result basically?

Reply:
> Agreed, but we'd still need to explicitly exclude fieldsFromSource, right?

Yeah, I guess what I'm getting at is: If `remainingFields` is determined by comparing `uniqFieldNames` with script_fields and runtime_mappings, and `uniqFieldNames` includes `fields` and `fieldsFromSource`, we could probably do this filtering a bit earlier on in one go by filtering `fields`, `fieldsFromSource`, `script_fields`, and `runtime_mappings`. Then you wouldn't have to worry about using any of those items in the rest of the function body because you'd know they are already filtered.

Not a huge concern or anything, just felt like it might make this method (marginally) more readable.